ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Mempool/Update.hs

```diff
@@ -172,28 +168,84 @@ pureTryAddTx ::
+         tx
+         err
+         (isMempoolSize is)
+        )
+    Right txsz
+      -- Check for overflow
+      --
+      -- No measure of a transaction can ever be negative, so the only way
+      -- adding two measures could result in a smaller measure is if some
+      -- modular arithmetic overflowed. Also, overflow necessarily yields a
+      -- lesser result, since adding 'maxBound' is modularly equivalent to
+      -- subtracting one. Recall that we're checking each individual addition.
+      --
+      -- We assume that the 'txMeasure' limit and the mempool capacity
+      -- 'isCapacity' are much smaller than the modulus, and so this should
+      -- never happen. Despite that, blocking until adding the transaction
+      -- doesn't overflow seems like a reasonable way to handle this case.
+     | not $ currentSize Measure.<= currentSize `Measure.plus` txsz
+     ->
+       NotEnoughSpaceLeft
+      -- We add the transaction if and only if it wouldn't overrun any component
+      -- of the mempool capacity.
+      --
+      -- In the past, this condition was instead @TxSeq.toSize (isTxs is) <
+      -- isCapacity is@. Thus the effective capacity of the mempool was
+      -- actually one increment less than the reported capacity plus one
+      -- transaction. That subtlety's cost paid for two benefits.
+      --
+      -- First, the absence of addition avoids a risk of overflow, since the
+      -- transaction's sizes (eg ExUnits) have not yet been bounded by
+      -- validation (which presumably enforces a low enough bound that any
+      -- reasonably-sized mempool would never overflow the representation's
+      -- 'maxBound').
+      --
+      -- Second, it is more fair, since it does not depend on the transaction
+      -- at all. EG a large transaction might struggle to win the race against
+      -- a firehose of tiny transactions.
+      --
+      -- However, we prefer to avoid the subtlety. Overflow is handled by the
+      -- previous guard. And fairness is already ensured elsewhere (the 'MVar's
+      -- in 'implAddTx' --- which the "Test.Consensus.Mempool.Fairness" test
+      -- exercises). Moreover, the notion of "is under capacity" becomes
+      -- difficult to assess independently of the pending tx when the measure
+      -- is multi-dimensional; both typical options (any component is not full
+      -- or every component is not full) lead to some confusing behaviors
+      -- (denying some txs that would "obviously" fit and accepting some txs
+      -- that "obviously" don't, respectively).
+      --
+      -- Even with the overflow handler, it's important that 'txMeasure'
+      -- returns a well-bounded result. Otherwise, if an adversarial tx arrived
+      -- that could't even fit in an empty mempool, then that thread would
+      -- never release the 'MVar'. In particular, we tacitly assume here that a
+      -- tx that wouldn't even fit in an empty mempool would be rejected by
+      -- 'txMeasure'.
```

Comment:
Also mention this in the haddocks of `txMeasure`?

Reply:
Great suggestion. I added an INVARIANT.